analytics-ios-mcvid/Classes/SEGMCVIDTracker.m

```diff
@@ -19,204 +25,348 @@ - (id)initWithCode:(MCVIDAdobeErrorCode)code message:(NSString *)message error:(
+    }
+    components.queryItems = queryItems;
+    NSURL *url = components.URL;
 
-  if ([context.payload isKindOfClass:[SEGScreenPayload class]]){
-    advertisingId = screen.context[@"device"][@"advertistingId"];
-  }
+    return url;
+}
 
-  if ([context.payload isKindOfClass:[SEGGroupPayload class]]){
-    advertisingId = group.context[@"device"][@"advertistingId"];
-  }
 
-  if (!self.staticMarketingCloudId) {
-    [self sendRequestAdobeExperienceCloud:advertisingId organizationId:organizationId completion:^(NSString *marketingCloudId, NSError *error) {
-      self.staticMarketingCloudId = marketingCloudId;
-    }];
+- (void)context:(SEGContext *_Nonnull)context next:(SEGMiddlewareNext _Nonnull)next {
+    if ([context.payload isKindOfClass:[SEGAliasPayload class]]) {
+      next(context);
+      return;
+    }
+
+  if (!self.cachedMarketingCloudId) {
+    next(context);
+    return;
   }
 
-  if (self.staticMarketingCloudId.length) {
-    NSMutableDictionary *mergedIntegrations = [NSMutableDictionary dictionaryWithCapacity:(identify.integrations.count || track.integrations.count || screen.integrations.count || group.integrations.count) + 1 ];
-    NSDictionary *mcidIntegrations = @{@"Adobe Analytics" : @{ @"marketingCloudVisitorId": self.staticMarketingCloudId } };
-    [mergedIntegrations addEntriesFromDictionary:mcidIntegrations];
+  SEGIdentifyPayload *identify =(SEGIdentifyPayload *)context.payload;
+  SEGTrackPayload *track =(SEGTrackPayload *)context.payload;
+  SEGScreenPayload *screen =(SEGScreenPayload *)context.payload;
+  SEGGroupPayload *group =(SEGGroupPayload *)context.payload;
+
+  if (self.cachedMarketingCloudId.length) {
+    NSMutableDictionary *adobeOptions = [NSMutableDictionary new];
 
     if ([context.payload isKindOfClass:[SEGIdentifyPayload class]]){
-      [mergedIntegrations addEntriesFromDictionary:identify.integrations];
-      SEGContext *newIdentifyContext = [context modify:^(id<SEGMutableContext> _Nonnull ctx) {
-          ctx.payload = [[SEGIdentifyPayload alloc] initWithUserId:identify.userId
+        if ((identify.integrations != NULL) && ([identify.integrations objectForKey:@"Adobe Analytics"])){
+            NSDictionary *existingAAOptions = identify.integrations[@"Adobe Analytics"];
+            [adobeOptions addEntriesFromDictionary:existingAAOptions];
+            [adobeOptions setObject:self.cachedMarketingCloudId forKey:@"marketingCloudVisitorId"];
+        }
+        
+        NSMutableDictionary *integrations = [NSMutableDictionary new];
+        if (identify.integrations != NULL) {
+            NSMutableDictionary *existingIntegrations = [NSMutableDictionary new];
+            [existingIntegrations addEntriesFromDictionary:identify.integrations];
+            [existingIntegrations removeObjectForKey:@"Adobe Analytics"];
+            [integrations addEntriesFromDictionary:existingIntegrations];
+            [integrations setObject:adobeOptions forKey:@"Adobe Analytics"];
+        }
```

Comment:
🚨 We have another serious bug here: if the payload's integrations dictionary is `nil` we won't send the `marketingCloudVisitorId`. The same applies to all the other types of payloads below.

Reply:
Also please, let's try to move this code into a single place instead of copying it into every kind of payload. I also feel like we could simplify this code a lot.